lib/hw_id/hw_id.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <hw_id.h>
+#include <zephyr/kernel.h>
+
+#define IMEI_LEN 15
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC)
+/* Ask BLE stack for default MAC address */
+
+#include <bluetooth/bluetooth.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+	struct bt_le_oob oob;
+	int ret = bt_le_oob_get_local(BT_ID_DEFAULT, &oob);
+
+	if (ret) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X",
+		oob.addr.a.val[5],
+		oob.addr.a.val[4],
+		oob.addr.a.val[3],
+		oob.addr.a.val[2],
+		oob.addr.a.val[1],
+		oob.addr.a.val[0]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID)
+/* Directly read Device ID from registers */
+
+#include <zephyr/drivers/hwinfo.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	uint8_t device_id[8];
+	ssize_t ret = hwinfo_get_device_id(device_id, ARRAY_SIZE(device_id));
+
+	if (ret != ARRAY_SIZE(device_id)) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X%02X%02X",
+		device_id[0],
+		device_id[1],
+		device_id[2],
+		device_id[3],
+		device_id[4],
+		device_id[5],
+		device_id[6],
+		device_id[7]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID)
+/* Request a UUID from the modem */
+
+#include <modem/modem_jwt.h>
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	struct nrf_device_uuid dev = {0};
+
+	int err = modem_jwt_get_uuids(&dev, NULL);
+
+	if (err) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%s", dev.str);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_IMEI)
+/* Ask the modem for an IMEI */
+
+#include <nrf_modem_at.h>
```

Comment:
I'm not the biggest fan of having includes spread throughout the source file. Its more difficult to identify the dependencies of the file.

Reply:
This would require having those `#if defined` statements copied to the top - not sure if this is prettier.

Reply:
I do see the point in having that as a convention though - will apply that change.